tutorial/lesson_21_auto_scheduler_generate.cpp

```diff
@@ -0,0 +1,190 @@
+    }
+
+    void schedule() {
+        if (!auto_schedule) {
+            return;
+        }
+
+        // To use the auto-scheduler, we need to provide estimates on the
+        // input/output sizes including estimates on the parameter values.
+
+        // First, provide estimates (min and extent values) for each dimension
+        // of the input images ('input', 'filter', and 'bias') using the
+        // set_bounds_estimate() method. set_bounds_estimate() takes in
+        // (min, extent) of the corresponding dimension as arguments.
+        input.dim(0).set_bounds_estimate(0, 131);
+        input.dim(1).set_bounds_estimate(0, 131);
+        input.dim(2).set_bounds_estimate(0, 64);
+        input.dim(3).set_bounds_estimate(0, 4);
+
+        filter.dim(0).set_bounds_estimate(0, 3);
+        filter.dim(1).set_bounds_estimate(0, 3);
+        filter.dim(2).set_bounds_estimate(0, 64);
+        filter.dim(3).set_bounds_estimate(0, 64);
+
+        bias.dim(0).set_bounds_estimate(0, 64);
+
+        // Next, provide estimates on the parameter values using the
+        // set_estimate() method.
+        min_value.set_estimate(2.0f);
+
+        // Last, provide estimates (min and extent values) for each dimension
+        // of pipeline outputs using the estimate() method. estimate() takes in
+        // (dim_name, min, extent) as arguments.
+        output1.estimate(x, 0, 128)
+            .estimate(y, 0, 128)
+            .estimate(z, 0, 64)
+            .estimate(n, 0, 4);
+
+        output2.estimate(x, 0, 128)
+            .estimate(y, 0, 128)
+            .estimate(z, 0, 64)
+            .estimate(n, 0, 4);
+
+        // Technically, the estimate values can be anything, but the closer
+        // they are to the actual use-case values, the better the generated
+        // schedule will be.
+
+        // Now, let's auto-schedule the pipeline by calling auto_schedule_outputs(),
+        // which takes in a MachineParams object as an argument. The machine_params
+        // argument is optional. If none is specified, the default machine parameters
+        // for generic CPU architecture are going to be used by the auto-scheduler.
+
+        // Let's use some random values for the machine parameters.
+        MachineParams machine_params(32, 16 * 1024 * 1024, 40);
+        // The arguments to MachineParams are the maximum level of parallelism
+        // available, the size of the last-level cache (in KB), and the ratio
+        // between memory cost and arithmetic cost at the last level case, of
+        // the target architecture in that order.
+
+        // Note that when using the auto-scheduler, no schedule should have
+        // been applied to the pipeline; otherwise, the auto-scheduler will
+        // throw an error. The current auto-scheduler does not work with
+        // partially-scheduled pipeline.
+        std::string schedule = auto_schedule_outputs(machine_params);
+        std::cout << "\nSchedule:\n\n" << schedule << "\n";
```

Comment:
I think we need to expand a bit on what happens here; if I didn't already know, it would not be clear to me what to do. (i.e.: the schedule that is dumped to count is actual Halide C++ source, which you are expected to copy-and-paste back into this very same source file.)